bindings/rust/s2n-tls/src/connection.rs

```diff
@@ -318,6 +318,45 @@ impl Connection {
         Ok(self)
     }
 
+    /// Sets the callback to use for verifying that a hostname from an X.509 certificate is
+    /// trusted.
+    ///
+    /// The callback may be called more than once during certificate validation as each SAN on
+    /// the certificate will be checked.
+    ///
+    /// Corresponds to the underlying
+    /// [C API](https://aws.github.io/s2n-tls/doxygen/s2n_8h.html#ad39f5c65f5c201d7af140baff29a02eb)
+    pub fn set_verify_host_callback<T: 'static + VerifyHostNameCallback>(
+        &mut self,
+        handler: T,
+    ) -> Result<&mut Self, Error> {
+        unsafe extern "C" fn verify_host_cb_fn(
+            host_name: *const ::libc::c_char,
+            host_name_len: usize,
+            context: *mut ::libc::c_void,
+        ) -> u8 {
+            let host_name = host_name as *const u8;
+            let host_name = core::slice::from_raw_parts(host_name, host_name_len);
+            if let Ok(host_name_str) = core::str::from_utf8(host_name) {
+                let context = &mut *(context as *mut Context);
+                let handler = context.verify_host_callback.as_mut().unwrap();
+                return handler.verify_host_name(host_name_str) as u8;
+            }
+            0 // If the host name can't be parsed, fail closed.
+        }
```

Comment:
Nope that will lead to using different `Context` types which I almost did https://github.com/aws/s2n-tls/pull/3925/commits/fba3ebe386b081927c370178ad819b52e86880cf

Reply:
The context is only necessary to get the handler though. The more interesting part of the method involves handling the host name and return value. It shouldn't lead to complicated looking code.

Reply:
While its there is some duplication, its only twice and being unsafe code, I think we should keep the entire callback function together to avoid loss of context.

For example, `from_raw_parts` is unsafe and moving that into its own function will make reasoning about it more difficult in the future.

Reply:
I don't agree. Encapsulating the common logic shouldn't lead to a loss of context if done correctly, and shouldn't make it any more unsafe. And right now the bigger maintainability danger is that the same callback can be called differently depending on how it's set, leading to very unexpected behavior.

But that's not happening at this moment, so this isn't a hard blocker if you really don't want to try to reduce the duplciation.